src/helpers/buffers/fsv.rs

```diff
@@ -28,81 +28,61 @@ use std::ops::Range;
 pub struct FixedSizeByteVec<const N: usize> {
     data: Vec<u8>,
     added: BitVec,
-    elements_drained: usize,
-    region_size: usize,
+    taken: usize,
 }
 
 impl<const N: usize> FixedSizeByteVec<N> {
     pub const ELEMENT_SIZE_BYTES: usize = N;
 
-    pub fn new(region_count: usize, region_size: usize) -> Self {
+    pub fn new(capacity: usize) -> Self {
         Self {
-            data: vec![0_u8; N * region_size * region_count],
-            added: bitvec![0; region_size * region_count],
-            elements_drained: 0,
-            region_size,
+            data: vec![0_u8; N * capacity],
+            added: bitvec![0; capacity],
+            taken: 0,
         }
     }
 
     /// Inserts a new element to the specified position, returning the previous element at this `index`.
     /// ## Panics
-    /// * Panics if `index` is out of bounds.
-    pub fn insert(&mut self, index: usize, elem: [u8; N]) -> Option<[u8; N]> {
-        let offset = Self::index_offset(index);
-
+    /// Panics if `index` is out of bounds or if something was previously inserted at `index`.
+    pub fn insert(&mut self, index: usize, elem: &[u8; N]) {
         // if index is out of bounds, this line will panic, there is no need for additional check
-        if self.added[index] {
-            // this is not supposed to be on the hot path, this branch is executed when there is
-            // an error and message with the same record id is received more than once, so
-            // copying here is fine.
-            let r = self.data[offset.clone()].try_into().unwrap();
-            self.data[offset].copy_from_slice(&elem);
-
-            Some(r)
-        } else {
-            self.added.set(index, true);
-            self.data[offset].copy_from_slice(&elem);
-
-            None
-        }
+        assert!(!self.added[index]);
```

Comment:
would `debug_assert` be appropriate here?

Reply:
I'm on the fence about this one.  Downgrading is work, so I'm going to say that we leave it for now.  It's a little annoying to find all the `#[should_panic]` tests and add `#[cfg(debug_assertions)]`.  I will add a TODO though.